### DISCUSSION TITLE
Generating a completely unique UUID per each object instantiation

### DISCUSSION THREAD

**Discussion by renardeinside** [pydantic/pydantic#8554] [Question]
Hi community:

Given this code, I'm running into quite a lot of inconveniences:

```py
from uuid import uuid4

from pydantic import BaseModel, Field, field_validator


class D(BaseModel):
    element_id: str | None = Field(default_factory=lambda: f"schorle-{uuid4()}")


class P(BaseModel):
    d: D = D()


p = P()
print(p.d.element_id)
p2 = P()
print(p2.d.element_id)
```

As you can imagine, the `element_id` is the same:
```
>>> schorle-4e9701c9-a323-46f3-b933-b8265e04626e
>>> schorle-4e9701c9-a323-46f3-b933-b8265e04626e
```
Even though parent object `P` instances are two different instances. 

Any idea how to fix such behaviour?

**Comment by uriyyo** [comment]
Hi @renardeinside,

You need to replace `P.d` field definition to use `default_factory` instead of default value:
```py
from uuid import uuid4

from pydantic import BaseModel, Field


class D(BaseModel):
    element_id: str | None = Field(default_factory=lambda: f"schorle-{uuid4()}")


class P(BaseModel):
    d: D = Field(default_factory=D)


p = P()
print(p.d.element_id)
p2 = P()
print(p2.d.element_id)
```
```
schorle-31f6a2ba-0040-4acf-a15c-daee8f13da10
schorle-0ff0bab7-b581-41ec-946a-7250d293d689
```

**Reply by renardeinside** [reply]
This indeed works, but quite alogical (at least from my perspective). 

I'm building a library based on pydantic, and this behavior seems to me like a leaky abstraction for the end user (it's an implementation detail, in the end). 
Is there any way to change this behavior?

**Reply by NEGU93** [reply]
I agree! It is quite anti intuitive. However, you solved my issue, thank you!

**Reply by GenhaoLi** [reply]
@renardeinside Late to the discussion, but I think it would be more confusing if `d: D = D()` is an implicit default factory. 

Suppose one needs to create a class instance and make it a constant default value for `P.d`, which is what your original code actually does and is a fairly common practice.

**Comment by robomotic** [comment]
And then you would also need a proper validator for your prefixed-UUID4 which is not a generic string?
Should one also implement a custom Field model?
